### PR TITLE
Expose Verifier in API, remove Linux webpki-roots dependency.

### LINKED ISSUE #7: expose the Verifier struct
Hi, thanks for the great work.

I am wondering if the Verifier struct be made public, I want to configure my own `ClientConfig`, for example, customize preferred cipher suites.

### LINKED ISSUE #12: Make Linux's webpki-roots optional and opt-in at runtime
Currently, the use of `webpki-roots` in the [Linux verifier](https://github.com/rustls/rustls-platform-verifier/blob/main/src/verification/others.rs#L70-L78) is not configurable and they are always loaded to supplement what might come from on-disk CA bundles. This, and the previous behavior of falling back to `webpki-roots`, was implemented at 1Password because we found `openssl-probe` to be way too unreliable and had multiple users write in that they were unable to use the app as a result.

However, this is likely not what everyone wants and makes it harder to control exactly what trust roots a consumer of `rustls-platform-verifier` might allow in addition to some other concerns:

> I think webpki-roots should not be included on Linux, since they will inevitably bitrot - @Ralith

But there we're also some solutions presented as well:

> I have mixed feelings about a feature, because those can end up silently enabled, which is a bit of a security hazard
maybe support manually adding certs and let people pipe in webpki-roots themselves if they really want to? - @Ralith

> The use of webpki-roots can be an optional feature enabled with explicit API - @djc

Out of these, the last idea to enable it with an explicit API seems to be the best balance between convenience, a simple crate API, and control/security. So I believe that is the way we should go.

There are then two things needed to resolve this:
- Make all `webpki-roots` dependencies optional via an opt-in feature called `webpki-roots` or `webpki-roots-augment`.
- Add an API to the top-level crate like this which sets a global atomic and then update `others.rs` to use it on Linux:
```rust
pub fn enable_webpki_fall() {
   ...
}

### PR DESCRIPTION
**Pull request by cpu**
## Description

This branch builds on https://github.com/rustls/rustls-platform-verifier/pull/23 and so should be reviewed commit-by-commit starting at b4db665fec0f24382bbfb9be6bbe6053dc77bf0c - the commits prior to this SHA are from #23.

The set of changes in this branch aim to accomplish two related goals:

1. Making the per-platform `Verifier` struct public, resolving https://github.com/rustls/rustls-platform-verifier/issues/7
2. Removing the Linux platform dependency on `webpki-roots`, replacing it with a new `Verifier::new_with_extra_roots` constructor, resolving https://github.com/rustls/rustls-platform-verifier/issues/12.
Callers can choose to use this constructor to provide roots from `webpki-roots` (or another source) and they will be added to the constructed `RootCertStore` in addition to trust anchors loaded from `rustls-native-certs`.

The `wasm32` target continues to depend on `webpki-roots` since there is no `rustls-native-certs` support for that target.

Resolves #7.
Resolves #12.

### TODO:

* [x] Land #23
* [x] Rebase

### CODE REVIEW

**Comment by cpu** [comment]
> cpu marked this pull request as draft now

Doing this for two reasons:
1. To focus review on #23 first.
2. To confirm this approach matches what @complexspaces et. al. were thinking in #12.

**Comment by djc** [comment]
For stacked branches, consider setting the target branch to the parent. That way, you don't see the extra commits and GitHub will change the target after the parent branch is merged.

File: Cargo.toml
Comment (Ralith) [inline Cargo.toml:41]: This seems like a small, portable API. Why bother masking it behind a feature?

**Comment by cpu** [comment]
> For stacked branches, consider setting the target branch to the parent. That way, you don't see the extra commits and GitHub will change the target after the parent branch is merged.

I think that only works if you push the parent branch to the origin repo, otherwise it will open a PR on my fork and not ping any reviewers. I'm not generally in the habit of pushing branches to the primary repo (they can end up collecting into messy clutter, and it's only possible with maintainer access) but I could consider it for these stacked changes.

**Review by Ralith — APPROVED** [review]
Seems like a great solution!

File: Cargo.toml
Comment (Ralith) [reply Cargo.toml:41]: I understand djc to be thinking of a webpki-specific API, but I think the less opinionated approach you have here is preferable in part *because* there's no need to gate it.

**Comment by Ralith** [comment]
Alternatively, rustls(?) could provide a wrapper for any `ServerCertVerifier` that adds additional trust roots, by checking against them before delegating to the inner verifier. That'd be a bit more general in principle, though less discoverable and perhaps not widely useful.

File: src/verification/android.rs
Comment (complexspaces) [inline src/verification/android.rs:71]: Suggestion: I believe this can be made `pub(crate)` like the other platforms. This is a leftover from when the tests which used this were outside of `src`.

File: src/verification/apple.rs
Comment (complexspaces) [inline src/verification/apple.rs:40]: I don't believe this is correct, was this a copy/paste error from the `other` verifier? We use both the system's root store and Apple's TLS verification routines that come from `Security.framework`. This one should be a closer match to the comment for Windows.

File: src/lib.rs
Comment (complexspaces) [inline src/lib.rs:49]: Nit: Maybe leave an extra comment at the top of the documentation block that mentions it uses safe defaults so people have an example of why they might need tighter control.

File: Cargo.toml
Comment (complexspaces) [reply Cargo.toml:41]: I like the idea of always exposing it. That would also match up better with [the API in `rustls` itself](https://docs.rs/rustls/latest/rustls/struct.ConfigBuilder.html#method.with_root_certificates) which is always available even without the dangerous configuration feature enabled.

**Comment by djc** [comment]
I was thinking in terms of a specific API call that makes it trivial to add the webpki-roots specifically, but maybe the we can make that easy enough anyway by way of the pki-types project?

File: Cargo.toml
Comment (djc) [reply Cargo.toml:41]: I generally wouldn't mask extra API behind a feature unless it requires extra dependencies or there is another motivation (like for dangerous configuration). I agree with @Ralith that it probably doesn't make sense here.

File: src/verification/others.rs
Comment (complexspaces) [inline src/verification/others.rs:100]: Nit: It would be nice to not need to duplicate these roots twice in memory for the lifetime of the verifier (three times if they're also hardcoded in the binary like `webpki-roots`. I'm not immediately sure to do this though since at this time we only have immutable access.

**Review by complexspaces — CHANGES_REQUESTED** [review]
I like how this is looking so far, thank you for picking this up.

File: Cargo.toml
Comment (complexspaces) [reply Cargo.toml:41]: GitHub put my comment elsewhere by mistake, but I'm also in favor of this change. It also lines up with the `rustls` API for adding arbitrary roots to a config not being gated.

File: src/verification/others.rs
Comment (Ralith) [inline src/verification/others.rs:45]: Would it be better to use `impl IntoIterator` here?

File: src/verification/apple.rs
Comment (cpu) [reply src/verification/apple.rs:40]: Copy-pasta err :spaghetti: :x: Thanks for catching.

File: src/verification/others.rs
Comment (cpu) [reply src/verification/others.rs:100]: I'll see if I can figure something out.

File: Cargo.toml
Comment (cpu) [reply Cargo.toml]: Removed the feature flag. Thanks for the input all :+1:

File: src/verification/android.rs
Comment (cpu) [reply src/verification/android.rs]: Tacked on an additional commit to implement this change.

File: src/verification/apple.rs
Comment (cpu) [reply src/verification/apple.rs]: Fixed up.

File: src/lib.rs
Comment (cpu) [reply src/lib.rs:50]: Added some text to emphasize that `tls_config` uses safe defaults.

File: src/verification/others.rs
Comment (cpu) [reply src/verification/others.rs]: Switched, thanks!

**Comment by cpu** [comment]
> I was thinking in terms of a specific API call that makes it trivial to add the webpki-roots specifically, but maybe the we can make that easy enough anyway by way of the pki-types project?

That seems like a good direction. It felt like for the purposes of platform-verifier it was better to be agnostic of the `webpki-roots` crate and just provide an option for an iterator of `OwnedTrustAnchors` of any provenance.

File: src/verification/others.rs
Comment (cpu) [reply src/verification/others.rs]: I'm not sure I see a path forward here :-/ Perhaps with some adjustments on the rustls side to not require owned representations of trust anchors?

@complexspaces Do you think this is blocking for the rest of the work?

Comment (complexspaces) [reply src/verification/others.rs]: I was thinking about this a little more, but couldn't we just wrap `extra_roots` in a `Mutex` and then use `Vec.drain` here? This shouldn't be in any kind of hot path since `init_verifier` is only called once per `Verifier` construction.

Comment (cpu) [reply src/verification/others.rs]: > couldn't we just wrap extra_roots in a Mutex and then use Vec.drain here?

That worked great, thank you. I had tried using a `Mutex` previously but I think I confused myself trying to use `get_mut` instead of `try_lock` and bumping into the same requirement of `mut self` (silly in retrospect).

I've used an `if let Ok(mut extra_roots) = self.extra_roots.try_lock() { ... }` construction in my last push. Do you think that's reasonable or should we handle the `Err` case differently? Maybe failing initialization?

Comment (complexspaces) [reply src/verification/others.rs]: Personally I would have just used `.unwrap()` because there's no way for the Mutex to get locked multiple times, so it could never be poisoned (which is the only route to a panic) but I'm fine with `if let Ok()` if others prefer that as long as we have a comment saying it can never fail.

Comment (cpu) [reply src/verification/others.rs]: > Personally I would have just used .unwrap() because there's no way for the Mutex to get locked multiple times, so it could never be poisoned (which is the only route to a panic)

That makes sense to me. I'll switch to `unwrap` and add a comment.

**Review by complexspaces — APPROVED** [review]

**Comment by cpu** [comment]
@djc Would you like me to hold this for you to do another pass, or can I merge?

**Comment by complexspaces** [comment]
FWIW the only other thing that came to mind is maybe allowing arbitrary roots to be added to any verifier for API consistency, but this is significantly complicated then adding them to the `webpki`-based verifier. Given that its probably best to wait for someone to present a use case  or PR for it.

**Comment by djc** [comment]
I don't need to do another pass.

**Pull request merged**
